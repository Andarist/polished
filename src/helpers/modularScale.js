// @flow
import stripUnit from './stripUnit'
import validateModule, { customRule } from '../validation/_validateModule'

const ratioNames = {
  minorSecond: 1.067,
  majorSecond: 1.125,
  minorThird: 1.2,
  majorThird: 1.25,
  perfectFourth: 1.333,
  augFourth: 1.414,
  perfectFifth: 1.5,
  minorSixth: 1.6,
  goldenSection: 1.618,
  majorSixth: 1.667,
  minorSeventh: 1.778,
  majorSeventh: 1.875,
  octave: 2,
  majorTenth: 2.5,
  majorEleventh: 2.667,
  majorTwelfth: 3,
  doubleOctave: 4,
}

/** */
type Ratio =
  | number
  | 'minorSecond'
  | 'majorSecond'
  | 'minorThird'
  | 'majorThird'
  | 'perfectFourth'
  | 'augFourth'
  | 'perfectFifth'
  | 'minorSixth'
  | 'goldenSection'
  | 'majorSixth'
  | 'minorSeventh'
  | 'majorSeventh'
  | 'octave'
  | 'majorTenth'
  | 'majorEleventh'
  | 'majorTwelfth'
  | 'doubleOctave'

/**
 * Establish consistent measurements and spacial relationships throughout your projects by incrementing up or down a defined scale. We provide a list of commonly used scales as pre-defined variables, see below.
 * @example
 * // Styles as object usage
 * const styles = {
 *    // Increment two steps up the default scale
 *   'fontSize': modularScale(2)
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *    // Increment two steps up the default scale
 *   fontSize: ${modularScale(2)}
 * `
 *
 * // CSS in JS Output
 *
 * element {
 *   'fontSize': '1.77689em'
 * }
 */
function modularScale(
  steps: number,
  base?: number | string = '1em',
  ratio?: Ratio = 'perfectFourth',
) {
  const realBase = typeof base === 'string' ? stripUnit(base) : base
  const realRatio = typeof ratio === 'string' ? ratioNames[ratio] : ratio

  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      !customRule('helpers/modularScale', {
        enforce: typeof realBase === 'number',
        msg: `expects base to be a valid em-based string value. However, you passed ${base} instead.`,
      })
    ) {
      return ''
    }
  }

  return `${realBase * realRatio ** steps}em`
}

export { ratioNames }
export default validateModule({
  modulePath: 'helpers/modularScale',
  types: [
    { type: 'number', required: true },
    { type: ['number', 'string'] },
    { type: ['number', 'enumberable'], map: ratioNames },
  ],
  errReturn: '',
})(modularScale)
