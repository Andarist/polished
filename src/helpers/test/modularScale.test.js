// @flow
import modularScale, { ratioNames } from '../modularScale'

describe('modularScale', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
      log: global.console.log,
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('should use perfectFourth and 1em base by default', () => {
    expect({ 'font-size': modularScale(1) }).toMatchSnapshot()
    expect({ 'font-size': modularScale(2) }).toMatchSnapshot()
    expect({ 'font-size': modularScale(0) }).toMatchSnapshot()
  })

  it('should allow adjusting the base', () => {
    expect({ 'font-size': modularScale(1, '2em') }).toMatchSnapshot()
  })

  it('should allow adjusting the base with a number value', () => {
    expect({ 'font-size': modularScale(1, 2) }).toMatchSnapshot()
  })

  it('should allow adjusting the ratio', () => {
    expect({ 'font-size': modularScale(1, '1em', 1) }).toMatchSnapshot()
  })

  it('should allow any of the predefined ratio names', () => {
    Object.keys(ratioNames).forEach(ratio => {
      expect({
        'font-size': modularScale(1, '1em', ratioNames[ratio]),
      }).toMatchSnapshot()
    })
  })

  it('should throw a warning if more than 3 parameters are provided', () => {
    modularScale(1, '1em', 'perfectFourth', true)
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })

  it('should throw an error if a non-number value for steps is provided', () => {
    modularScale('invalid')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error if a non-number or non-string value for base is provided', () => {
    modularScale(2, true)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error if an invalid string-based value for base is provided', () => {
    modularScale(2, 'invalid')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error if an invalid ratio is provided', () => {
    modularScale(2, '1em', 'invalid')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })
})
