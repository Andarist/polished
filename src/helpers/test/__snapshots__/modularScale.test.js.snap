exports[`modularScale should allow adjusting the base 1`] = `
Object {
  "font-size": "2.666em",
}
`;

exports[`modularScale should allow adjusting the base with a number value 1`] = `
Object {
  "font-size": "2.666em",
}
`;

exports[`modularScale should allow adjusting the ratio 1`] = `
Object {
  "font-size": "1em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 1`] = `
Object {
  "font-size": "1.067em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 2`] = `
Object {
  "font-size": "1.125em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 3`] = `
Object {
  "font-size": "1.2em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 4`] = `
Object {
  "font-size": "1.25em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 5`] = `
Object {
  "font-size": "1.333em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 6`] = `
Object {
  "font-size": "1.414em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 7`] = `
Object {
  "font-size": "1.5em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 8`] = `
Object {
  "font-size": "1.6em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 9`] = `
Object {
  "font-size": "1.618em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 10`] = `
Object {
  "font-size": "1.667em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 11`] = `
Object {
  "font-size": "1.778em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 12`] = `
Object {
  "font-size": "1.875em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 13`] = `
Object {
  "font-size": "2em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 14`] = `
Object {
  "font-size": "2.5em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 15`] = `
Object {
  "font-size": "2.667em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 16`] = `
Object {
  "font-size": "3em",
}
`;

exports[`modularScale should allow any of the predefined ratio names 17`] = `
Object {
  "font-size": "4em",
}
`;

exports[`modularScale should throw a warning if more than 3 parameters are provided 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- helpers/modularScale.js --%c

modularScale %cexpects a maximum of %c3 parameters%c. However, you passed %c4 parameters%c. This additional parameter was ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#modularScale%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`modularScale should throw an error if a non-number or non-string value for base is provided 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- helpers/modularScale.js --%c

modularScale %cexpects a 2nd parameter of type number,em. However, you passed true(boolean) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#modularScale%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`modularScale should throw an error if a non-number value for steps is provided 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- helpers/modularScale.js --%c

modularScale %cexpects a 1st parameter of type number. However, you passed invalid(string) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#modularScale%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`modularScale should throw an error if an invalid ratio is provided 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- helpers/modularScale.js --%c

modularScale %cexpects a 3rd parameter of type number,enumberable. However, you passed invalid(string) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#modularScale%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`modularScale should throw an error if an invalid string-based value for base is provided 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- helpers/modularScale.js --%c

modularScale %cexpects a 2nd parameter of type number,em. However, you passed invalid(string) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#modularScale%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`modularScale should use perfectFourth and 1em base by default 1`] = `
Object {
  "font-size": "1.333em",
}
`;

exports[`modularScale should use perfectFourth and 1em base by default 2`] = `
Object {
  "font-size": "1.776889em",
}
`;

exports[`modularScale should use perfectFourth and 1em base by default 3`] = `
Object {
  "font-size": "1em",
}
`;
